core/include/gnuradio-4.0/TriggerMatcher.hpp

```diff
@@ -0,0 +1,302 @@
+trim(std::string_view str) noexcept {
+    const auto first = str.find_first_not_of(" \t\n\r\f\v");
+    if (first == std::string_view::npos) return ""; // all whitespace
+
+    const auto last = str.find_last_not_of(" \t\n\r\f\v");
+    return str.substr(first, last - first + 1);
+}
+
+inline void
+parse(const std::string_view &trigger, std::string &triggerName, bool &triggerNameEnds, std::string &triggerCtx, bool &triggerCtxEnds) {
+    if (const size_t first_pos = trigger.find(SEPARATOR); first_pos != std::string::npos) {
+        if (trigger.find(SEPARATOR, first_pos + SEPARATOR.size()) != std::string::npos) {
+            throw gr::exception(fmt::format("invalid trigger input: multiple '{}' separators found: '{}'", SEPARATOR, trigger));
+        }
+
+        // separator found, split the trigger name and context
+        triggerName = trim(trigger.substr(0, first_pos));
+        triggerCtx  = trim(trigger.substr(first_pos + SEPARATOR.size()));
+        if (triggerNameEnds = triggerName.starts_with('^'); triggerNameEnds) {
+            triggerName = trim(triggerName.substr(1));
+        }
+        if (triggerCtxEnds = triggerCtx.starts_with('^'); triggerCtxEnds) {
+            triggerCtx = trim(triggerCtx.substr(1));
+        }
+    } else { // no separator, the whole string is considered as trigger name
+        triggerName = trim(trigger);
+        if (triggerNameEnds = triggerName.starts_with('^'); triggerNameEnds) {
+            triggerName = trim(triggerName.substr(1));
+        }
+        triggerCtx = "";
+    }
+};
+} // namespace detail
+
+namespace BasicTriggerNameCtxMatcher {
+
+void
+reset(property_map &state) noexcept {
+    state["triggerActive"]           = false;
+    state["waitingForStartNonMatch"] = false;
+    state["waitingForStopNonMatch"]  = false;
+}
+
+[[nodiscard]] bool
+isSingleTrigger(const property_map &state) noexcept {
+    if (state.contains("startDefined") && state.contains("stopDefined")) {
+        return std::get<bool>(state.at("startDefined")) xor std::get<bool>(state.at("stopDefined"));
+    } else {
+        return false;
+    }
+}
+
+void
+verifyFilterState(std::string_view matchCriteria, property_map &state) {
+    using namespace std::string_literals;
+    using namespace gr::trigger::detail;
+    if (state.contains("filter") && std::holds_alternative<std::string>(state.at("filter")) && (std::get<std::string>(state.at("filter")) == matchCriteria)) {
+        return;
+    }
+
+    state["filter"] = std::string(matchCriteria);
+
+    // default config
+    state["startDefined"]         = false;
+    state["stopDefined"]          = false;
+    state["startTriggerName"]     = ""s;
+    state["startCtx"]             = ""s;
+    state["stopTriggerName"]      = ""s;
+    state["stopCtx"]              = ""s;
+    state["startTriggerNameEnds"] = false;
+    state["startCtxEnds"]         = false;
+    state["stopTriggerNameEnds"]  = false;
+    state["stopCtxEnds"]          = false;
```

Comment:
Should we use strings constants for the keys?

Reply:
done... will push shortly.